### PR TITLE
Device Binding Callback implementation

### PR DESCRIPTION
**Pull request by vahancouver**
# JIRA Ticket

[SDKS-1748](https://bugster.forgerock.org/jira/browse/SDKS-1748)

# Description

Device Binding callback implementation
FRUI changes for testing in the example app

Note: tests will be implemented as a separate ticket

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        
+        guard let callbackType = json[CBConstants.type] as? String else {
+            throw AuthError.invalidCallbackResponse(String(describing: json))
+        }
+        
+        guard let outputs = json[CBConstants.output] as? [[String: Any]], let inputs = json[CBConstants.input] as? [[String: Any]] else {
+            throw AuthError.invalidCallbackResponse(String(describing: json))
+        }
+        
+        for output in outputs {
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:62]: `can we refactor this code ? we could do some thing like this

   inputs.forEach { input in
            foo(input[CBConstants.name] as? String)
        }
        
        `func` foo(name: String) {
        switch name {
        case Constants.jws: {
            self.jws = name
        }
        }
    }
`

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+//  FRAuth
+//
+//  Copyright (c) 2022 ForgeRock. All rights reserved.
+//
+//  This software may be modified and distributed under the terms
+//  of the MIT license. See the LICENSE file for details.
+//
+
+
+import Foundation
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:12]: Add new line separator

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+//  of the MIT license. See the LICENSE file for details.
+//
+
+import Security
+import Foundation
+import FRCore
+import JOSESwift
+
+
+/// Protiocol  to override keypair generation, authentication, signing and access control
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:17]: Typo issue

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    /// biometric handler for autentication
+    var biometricInterface: BiometricHandler
+    /// keyAware for key pair generation
+    var keyAware: KeyAware
+    
+    
+    /// Initializer for BiometricOnly
+    /// - Parameter biometricInterfcae: biometric handler for autentication
+    /// - Parameter keyAware: keyAware for key pair generation
+    init(biometricInterfcae: BiometricHandler, keyAware: KeyAware) {
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:95]: Typo issue. `biometricInterface`

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
not done yet..will review sometime later

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift (hunk 1 above)
Comment (rodrigoareis) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:62]: Do you think we can simplify this interaction logic?

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            else if let outputName = output[CBConstants.name] as? String, outputName == CBConstants.timeout, let outputValue = output[CBConstants.value] as? Int {
+                timeout = outputValue
+            }
+        }
+        
+        guard userId != nil, userName != nil, challenge != nil, deviceBindingAuthenticationType != nil, title != nil, subtitle != nil, promptDescription != nil else {
+            throw AuthError.invalidCallbackResponse("Missing an output value")
+        }
+        
+        for input in inputs {
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:93]: Do you think we can simplify this interaction logic?

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Left minor comments to be reviewed. Is there any kind of tests for new `DeviceBindingCallback`?

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift (hunk 1 above)
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:62]: +1

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    func bind(completion: @escaping DeviceBindingResultCallback) {
+        execute(authInterface: nil, deviceId: nil, completion)
+    }
+    
+    
+    /// Helper method to execute binding , signing, show biometric prompt.
+    /// - Parameter authInterface: Interface to find the Authentication Type - procide nil to default to getDeviceBindingAuthenticator()
+    /// - Parameter deviceId: Interface to find the Authentication Type - provide nil to default to FRDevice.currentDevice?.identifier.getIdentifier()
+    /// - Parameter completion: completion Completion block for Device binding result callback
+    open func execute(authInterface: DeviceAuthenticator?,
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:130]: i am not sure we need open,  protected open is a standard in kotlin to override and method is not exposed , in swift open method can be called by anyone i believe . so i am kind of thinking this

`class DeviceBindingCallback {
    fileprivate func execute() {
        print("Foo")
    }
}

class Bar: DeviceBindingCallback {
    fileprivate override func jey() {
        print("Bar")
    }
}`

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift (hunk 1 above)
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:62]: my idea is something like this. 

    inputs.forEach { input in
            foo(input[CBConstants.name] as? String)
        }

 `    func foo(name: String) {
        switch name {
        case Constants.jws: {
            self.jws = name
        }
       case Constants.authenticationType: {
            self.deviceBindingAuthenticationType = DeviceBindingAuthenticationType(rawValue: name)
        }
        }
    }`

File: FRAuth/FRAuth/DeviceBinding/BiometricBindingHandler.swift [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    func isSupported(policy: LAPolicy) -> Bool {
+        let laContext = LAContext()
+        var evalError: NSError?
+        return laContext.canEvaluatePolicy(policy, error: &evalError)
+    }
+    
+    
+    /// Display biometric prompt  for Biometric and device credential if needed
+    /// - Parameter timeout: Timeout for the biometric prompt
+    /// - Parameter completion: Completion block for Device binding result callback
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/BiometricBindingHandler.swift:55]: format this class , looks like its not formatted

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    /// Default implemention
+    /// Sign the challenge sent from the server and generate signed JWT
+    /// - Parameter keyPair: Public and private key pair
+    /// - Parameter kid: Generated key id
+    /// - Parameter userId: user Id received from server
+    /// - Parameter challenge: challenge received from server
+    /// - Returns: compact serialized jws
+    func sign(keyPair: KeyPair, kid: String, userId: String, challenge: String) throws -> String {
+        
+        let jwk = try RSAPublicKey(publicKey: keyPair.publicKey, additionalParameters: ["use": "sig", "alg": "RS512"])
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:55]: we need to create a keys using secure enclave ?

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                self.jwsKey = name
+            }
+            else if let name = input[CBConstants.name] as? String, name.contains(CBConstants.deviceName) {
+                self.deviceNameKey = name
+            }
+            else if let name = input[CBConstants.name] as? String, name.contains(CBConstants.deviceId) {
+                self.deviceIdKey = name
+            }
+            else if let name = input[CBConstants.name] as? String, name.contains(CBConstants.clientError) {
+                self.clientErrorKey = name
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:104]: For consistency, do we need self.jwsKey = name or just jwsKey = name, if the global variable and local variable same , people prefer self.

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+                self.handleException(status: .unsupported(errorMessage: error.localizedDescription), completion: completion)
+            }
+        }
+        
+    }
+    
+    
+    /// Handle all the errors for the device binding.
+    /// - Parameter status: Device binding status
+    /// - Parameter completion: Completion block Device binding result callback
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:170]: open should not be exposed, same comment as above

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    /// The userId received from server
+    private(set) var userId: String!
+    /// The userName received from server
+    private(set) var userName: String!
+    /// The challenge received from server
+    private(set) var challenge: String!
+    /// The authentication type of the journey
+    private(set) var deviceBindingAuthenticationType: DeviceBindingAuthenticationType!
+    // The title to be displayed in biometric prompt
+    private(set) var title: String!
```
Comment (george-bafaloukas-forgerock) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:29]: Should that be forced?
Can we not have them without the "!" either some as optional or the rest set with a default value or on the init

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+            
+        }
+    }
+}
+
+
+
+
+
+
```
Comment (george-bafaloukas-forgerock) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:266]: There are some formatting issues on this file, double spaces and extra lines

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+        
+        let newAuthInterface = authInterface ?? getDeviceBindingAuthenticator()
+        let newDeviceId = deviceId ?? FRDevice.currentDevice?.identifier.getIdentifier()
+        
+        guard newAuthInterface.isSupported() else {
+            handleException(status: .unsupported(errorMessage: nil), completion: completion)
+            return
+        }
+        
+        newAuthInterface.authenticate(timeout: timeout ?? 60) { result in
```
Comment (george-bafaloukas-forgerock) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:142]: Are we capturing self on this call? might be work changing that to 
"newAuthInterface.authenticate(timeout: timeout ?? 60) {  [unowned self] result in"

And test if we are OK

**Review by george-bafaloukas-forgerock — CHANGES_REQUESTED** [review]

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift (hunk 6 above)
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:130]: so i dont want this 

let callback = DeviceBindCallBack()
callback.execute()

but i want this

class GeorgeCallback: DeviceBindCallBack {
   override func execute()
}

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        
+        try super.init(json: json)
+        
+        self.type = callbackType
+        self.response = json
+    }
+    
+    
+    /// Bind the device.
+    /// - Parameter completion Completion block for Device binding result callback
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:120]: we want this to be open func bind()

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift (hunk 8 above)
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:55]: please check they support RSA/SHA and let us know . Andy may have to do some work on server if thats the only thing iOS supports

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+        let jwkWithKeyId = try  jwk.withThumbprintAsKeyId()
+        
+        //create header
+        var header = JWSHeader(algorithm: .RS512)
+        header.kid = kid
+        header.typ = "JWS"
+        header.jwkTyped = jwkWithKeyId
+        
+        //create payload
+        let params: [String: Any] = ["sub": userId, "challenge": challenge, "exp": (Int(Date().addingTimeInterval(30).timeIntervalSince1970))]
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:65]: The `exp` value should be set with the value from the `timeout` attribute coming with the callback. 
In android, the following logic is in place:

>   /**
>      * Get Expiration date for the signed token, claim "exp" will be set to the JWS.
>      *
>      * @return The expiration date
>      */
>     protected open fun getExpiration(): Date {
>         val date = Calendar.getInstance();
>         date.add(Calendar.SECOND, timeout ?: 60)
>         return date.time;
>     }

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+        let message = try JSONSerialization.data(withJSONObject: params, options: [])
+        let payload = Payload(message)
+        
+        //create signer
+        guard let signer = Signer(signingAlgorithm: .RS512, key: keyPair.privateKey) else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot creat a signer for jws")
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:71]: Typo in `errorMessage`: "creat" => "create"

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+        return jws.compactSerializedString
+    }
+    
+}
+
+
+/// DeviceAuthenticator adoption for  biometric only authentication
+internal struct BiometricOnly: DeviceAuthenticator {
+    
+    /// biometric handler for autentication
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:86]: Typo: "autentication" => "authentication"

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+        //create jws
+        let jws = try JWS(header: header, payload: payload, signer: signer)
+        
+        return jws.compactSerializedString
+    }
+    
+}
+
+
+/// DeviceAuthenticator adoption for  biometric only authentication
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:83]: Remove extra space between "for" and "biometric"

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+    /// Initializer for BiometricOnly
+    /// - Parameter biometricInterfcae: biometric handler for autentication
+    /// - Parameter keyAware: keyAware for key pair generation
+    init(biometricInterfcae: BiometricHandler, keyAware: KeyAware) {
+        self.biometricInterface = biometricInterfcae
+        self.keyAware = keyAware
+    }
+    
+    
+    /// Generate  public and private key pair
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:101]: Remove extra space between "Generate" and "public"

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+    func generateKeys() throws -> KeyPair {
+        var keyBuilderQuery = keyAware.keyBuilderQuery()
+        keyBuilderQuery[String(kSecAttrAccessControl)] = accessControl()
+        return try keyAware.createKeyPair(builderQuery: keyBuilderQuery) 
+    }
+    
+    
+    /// Display authentication  prompt for authentication type if needed
+    /// - Parameter timeout: Timeout for teh authentication prompt
```
Comment (spetrov) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:110]: Typo: "teh"...
Remove extra space in "authentication  prompt"

**Review by spetrov — CHANGES_REQUESTED** [review]
Reviewed. I have a few minor issues that we may need to address...

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    
+    
+    /// Query attributes dictionary for generating key pair
+    func keyBuilderQuery() -> [String: Any] {
+        var query = [String: Any]()
+        
+        query[String(kSecAttrKeyType)] = String(keyType)
+        query[String(kSecAttrKeySizeInBits)] = keySize
+
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:40]: i belive the secure enclave the part will be somewhere here 

` #if !targetEnvironment(simulator)
        // If the device supports Secure Enclave, create a keypair using Secure Enclave TokenID
        if SecuredKey.isAvailable() {
            query[String(kSecAttrTokenID)] = String(kSecAttrTokenIDSecureEnclave)
            let accessControl = SecAccessControlCreateWithFlags(kCFAllocatorDefault, accessibility.rawValue as CFString, [.privateKeyUsage, .userPresence] , nil)!
            keyAttr[String(kSecAttrAccessControl)] = accessControl
        }
        #endif`

**Comment by vahancouver** [comment]
@spetrov @jeyanthanperiyasamy @rodrigoareis @george-bafaloukas-forgerock Please have another look. 

I addressed all the comments related to formatting, double spaces and typos.

I have changed to use the Secure Enclave if available and the token is generated now using ES256 algorithm.

@george-bafaloukas-forgerock @jeyanthanperiyasamy let's connect early morning my time to go over the parsing code in DeviceBindingCallback.init method and the reason of using forced unwrapped properties. I made some changes here to use switch [though arguably I preferred the previous code as it was more readable]

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Cookie/CookieValidationTests.swift [hunk 22]
```diff
@@ -202,7 +202,7 @@ class CookieValidationTests: FRAuthBaseTest {
     
     
     func test_10_cookie_is_expired_validation_not_expired() {
-        let setCookie: [String: String] = ["Set-Cookie":"iPlanetDirectoryPro=token; Expires=Wed, 21 Oct 2022 01:00:00 GMT; Domain=openam.example.com"]
+        let setCookie: [String: String] = ["Set-Cookie":"iPlanetDirectoryPro=token; Expires=Wed, 21 Oct 2032 01:00:00 GMT; Domain=openam.example.com"]
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Cookie/CookieValidationTests.swift:205]: is this part of this story ? or its a accident commit

File: FRTestHost/FRTestHost/FRAuthConfigPKHash.plist [hunk 23]
```diff
@@ -26,7 +26,7 @@
 	<string>signUp</string>
 	<key>forgerock_ssl_pinning_public_key_hashes</key>
 	<array>
-		<string>+KSzREQbAh9gqYLLGpfCG+cAy7Px3/Qmk/e8Egwyd7o=</string>
+		<string>9uiEUVtJyf/LAciFcfkWlBcrPor9uZGmteP3x9sBGiQ=</string>
```
Comment (jeyanthanperiyasamy) [inline FRTestHost/FRTestHost/FRAuthConfigPKHash.plist:29]: is this part of this story ? or its a accident commit

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
most of the changes looks good now, Lets test the ECC encryption type , once and wrap this story

**Comment by vahancouver** [comment]
@jeyanthanperiyasamy @spetrov 
Turns out the system triggers (biometric) authentication on signing operation when a private key is used that was created in secure enclave with the appropriate access flags. 
Because of this for the BIOMETRIC_ONLY and BIOMETRIC_ALLOW_FALLBACK types we had double authentication (I was triggering the second one in the code explicitly). 
With the latest changes I removed the manual triggering of the authentication. Now it gets triggered only once by the system during the JWS signing operation. 
Please review my changes.

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+//
+
+
+import Foundation
+import CommonCrypto
+import FRCore
+
+///Helper struct to generate and sign the keys
+struct KeyAware {
+    var userId: String
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:18]: userid and time is suppose to be private set and public get ?

File: FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+import JOSESwift
+
+/**
+ * Callback to collect the device binding information
+ */
+open class DeviceBindingCallback: MultipleValuesCallback {
+    
+    //  MARK: - Properties
+    
+    /// The userId received from server
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/Callbacks/DeviceBindingCallback.swift:22]: can you check the Android changes for this , all this needs to be private set and public get i belive ?

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+    /// Check if authentication is supported
+    func isSupported() -> Bool {
+        let laContext = LAContext()
+        var evalError: NSError?
+        return laContext.canEvaluatePolicy(policy, error: &evalError)
+    }
+    
+    
+    /// Access Control for the authetication type
+    func accessControl() -> SecAccessControl? {
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:123]: lets discuss about this fallback mechanism

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 27]
```diff
@@ -0,0 +1,250 @@
...
+#if !targetEnvironment(simulator)
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:176]: same we need to discuss about this

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+        
+        query[String(kSecAttrKeyType)] = String(keyType)
+        query[String(kSecAttrKeySizeInBits)] = keySize
+        
+        var keyAttr = [String: Any]()
+        keyAttr[String(kSecAttrIsPermanent)] = true
+        keyAttr[String(kSecAttrApplicationTag)] = key
+        query[String(kSecPrivateKeyAttrs)] = keyAttr
+        
+#if !targetEnvironment(simulator)
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:45]: if the secure key is not available , should we fail or guide the users ? something to think of here

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
feedback given

**Review by spetrov — APPROVED** [review]

**Review by rodrigoareis — APPROVED** [review]
Changes looks good to me

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+        return SecAccessControlCreateWithFlags(kCFAllocatorDefault, kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly, SecuredKey.isAvailable() ? [.userPresence, .privateKeyUsage] : [.userPresence], nil)!
+#else
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:178]: i am thinking like if its debug mode use this  [.userPresence] and for release [.userPresence, .privateKeyUsage] ...so we can remove the SecureEnclave check ...thoughts ?

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    
+    
+    /// Creates Keypair for the given builder query attributes
+    /// - Parameter builderQuery: query attributes dictionary for generating key pair
+    /// - Throws: error during private/public key generation
+    func createKeyPair(builderQuery: [String: Any]) throws -> KeyPair {
+        
+        var error: Unmanaged<CFError>?
+        guard let privateKey = SecKeyCreateRandomKey(builderQuery as CFDictionary, &error) else {
+            throw error!.takeRetainedValue() as Error
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:63]: can this be optional   - error?.takeRetainedValue()

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+        query[String(kSecAttrKeyType)] = String(kSecAttrKeyTypeEC)
+        query[String(kSecReturnRef)] = true
+        query[String(kSecAttrApplicationTag)] = keyAlias
+        
+        var item: CFTypeRef?
+        let status = SecItemCopyMatching(query as CFDictionary, &item)
+        guard status == errSecSuccess else {
+            return nil
+        }
+        return (item as! SecKey)
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:90]: can this be optional - return (item as? SecKey)     ?

File: FRAuth/FRAuth/DeviceBinding/KeyAware.swift [hunk 32]
```diff
@@ -0,0 +1,114 @@
...
+        query[String(kSecClass)] = String(kSecClassKey)
+        query[String(kSecAttrApplicationTag)] = keyAlias
+        SecItemDelete(query as CFDictionary)
+    }
+    
+    /// Get hash for the given key name (user id)
+    /// - Parameter keyName: key names to be hashed
+    /// - Returns: the hash for the given key name
+    static func getKeyAlias(keyName: String) -> String {
+        let data = keyName.data(using: .utf8)!
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/KeyAware.swift:107]: same here...can we do without bangs

File: FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift [hunk 33]
```diff
@@ -0,0 +1,250 @@
...
+        var evalError: NSError?
+        return laContext.canEvaluatePolicy(policy, error: &evalError)
+    }
+    
+    
+    /// Access Control for the authetication type
+    func accessControl() -> SecAccessControl? {
+#if !targetEnvironment(simulator)
+        return SecAccessControlCreateWithFlags(kCFAllocatorDefault, kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly, SecuredKey.isAvailable() ? [.biometryCurrentSet, .privateKeyUsage] : [.biometryCurrentSet], nil)!
+#else
```
Comment (jeyanthanperiyasamy) [inline FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift:126]: do we need this check and bang - SecuredKey.isAvailable()

**Review by jeyanthanperiyasamy — CHANGES_REQUESTED** [review]
verify what comments we can address

**Review by jeyanthanperiyasamy — APPROVED** [review]

**Pull request merged**